### PR TITLE
CCurlFile Authentication fix

### LINKED ISSUE #386: Add support for the HTTP digest authentication (for image / icon requests)
https://tvheadend.org/issues/5370

### PR DESCRIPTION
**Pull request by ksooo**
If username and password are given but no authentication method, let CURL try to choose an appropriate authentication method (CURLAUTH_ANY). Fixes https://github.com/kodi-pvr/pvr.hts/issues/386 and https://tvheadend.org/issues/5370

Runtime tested an macOS and Android with latest kodi master, latest pvr.hts and latest tvheadend. Works just fine, no regressions found.

@phunkyfish when you find some time...

### CODE REVIEW

**Comment by ksooo** [comment]
Breaks some tests. Will check.

**Comment by ksooo** [comment]
Jenkins is happy now. 
@phunkyfish PR is ready for review.

File: xbmc/filesystem/CurlFile.cpp [hunk 1]
```diff
@@ -565,6 +560,19 @@ void CCurlFile::SetCommonOptions(CReadState* state, bool failOnError /* = true *
...
+    else
+      bAuthSet = false;
+  }
+
+  // set username and password for current handle
+  if (!m_username.empty())
+  {
+    g_curlInterface.easy_setopt(h, CURLOPT_USERNAME, m_username.c_str());
+    if (!m_password.empty())
+      g_curlInterface.easy_setopt(h, CURLOPT_PASSWORD, m_password.c_str());
```
Comment (phunkyfish) [inline xbmc/filesystem/CurlFile.cpp:572]: From looking at the docs when using these newer opts which separate username and password the values will be URL decoded. Therefore I think you should URL encode the values beforehand.

Ref: https://curl.se/libcurl/c/CURLOPT_USERPWD.html

**Review by phunkyfish — COMMENTED** [review]
With the second commit I will wonder will there be side effects with various add-ons dependent on the old incorrect behaviour. But I guess there is hard to say or figure out without getting this out in the wild.

Anyone else with a little more CURL expertise that could look at this?

File: xbmc/filesystem/CurlFile.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/filesystem/CurlFile.cpp:572]: Are you sure? I read the docs differently.

Comment (phunkyfish) [reply xbmc/filesystem/CurlFile.cpp:572]:
```
The user and password strings are not URL decoded, so there's no way to send in a user name containing a colon using this option. Use CURLOPT_USERNAME for that, or include it in the URL.
```

Comment (fuzzard) [reply xbmc/filesystem/CurlFile.cpp:572]: I read that as only relevant for the old style. There's no mention of encoding on the 2 new style pages at all.
As the new style is 2 separate calls (use and pass) there's no need for a distinction for the colon separated like the old style

Comment (ksooo) [reply xbmc/filesystem/CurlFile.cpp:572]: I had a quick look at the curl sources and found no trace of user/pass strings being decoded.

Comment (phunkyfish) [reply xbmc/filesystem/CurlFile.cpp:572]: Ok, fuzzard’s explanation makes sense. Sorry for the noise.

**Comment by ksooo** [comment]
@phunkyfish The second commit actually restores the behavior to what it was before the first commit. So, although there always can be any side effects, the chances for breaking something is nearly zero, given both commits get merged.

**Comment by ksooo** [comment]
Well, just did a deep dive onti our code. The second commit does not change behavior for addons. What might be different with this commit is the result of a call to ' g_curlInterface.easy_getinfo(..., CURLINFO_CONTENT_TYPE, ...)', but I cannot find that this gets called by the addon filesystem API implementation. Moreover, for obtaining content-type the filesystem API uses what this PR does. 

https://github.com/xbmc/xbmc/blob/master/xbmc/addons/interfaces/Filesystem.cpp#L591 => https://github.com/xbmc/xbmc/blob/master/xbmc/filesystem/CurlFile.cpp#L1890 =>  https://github.com/xbmc/xbmc/blob/master/xbmc/filesystem/CurlFile.cpp#L1997

From my point of view it's afe to merge this PR.

**Review by phunkyfish — APPROVED** [review]
Fair, your explanation makes perfect sense and brings consistency which is better.

**Comment by arnova** [comment]
@ksooo : Mind having a look at https://github.com/xbmc/xbmc/issues/22035 ? It seems to be caused by the changes in this PR. Also seems rather brutal to always try to perform an unauthenticated connection first for every single Curl call we do.

**Pull request merged**